packages/console/src/hooks/use-subscribe.ts

```diff
@@ -0,0 +1,79 @@
+import { nanoid } from 'nanoid';
+import { toast } from 'react-hot-toast';
+
+import { useCloudApi } from '@/cloud/hooks/use-cloud-api';
+import { type CreateTenantData } from '@/components/CreateTenantModal/type';
+import { getBasename } from '@/consts';
+import { checkoutStateQueryKey, checkoutSuccessCallbackPath } from '@/consts/subscriptions';
+import { createLocalCheckoutSession } from '@/utils/checkout';
+
+type SubscribeProps = {
+  planId: string;
+  callbackPage?: string;
+  tenantId?: string;
+  tenantData?: CreateTenantData;
+  isDowngrade?: boolean;
+};
+
+const useSubscribe = () => {
+  const cloudApi = useCloudApi();
+
+  const subscribe = async ({
+    planId,
+    callbackPage,
+    tenantId,
+    tenantData,
+    isDowngrade = false,
+  }: SubscribeProps) => {
+    const state = nanoid(6);
+
+    const successSearchParam = new URLSearchParams({
+      [checkoutStateQueryKey]: state,
+    });
+
+    const successCallbackUrl = new URL(
+      `${getBasename()}/${checkoutSuccessCallbackPath}?${successSearchParam.toString()}`,
+      window.location.origin
+    ).toString();
+
+    const { redirectUri, sessionId } = await cloudApi.post('/api/checkout-session', {
+      body: {
+        planId,
+        successCallbackUrl,
+        tenantId,
+        tenantName: tenantData?.name,
+        tenantTag: tenantData?.tag,
+      },
+    });
+
+    if (!redirectUri) {
+      // Note: This should never happen in theory
+      toast.error('Something went wrong, please try again later');
+      return;
+    }
+
+    createLocalCheckoutSession({
+      state,
+      sessionId,
+      callbackPage,
+      isDowngrade,
+    });
+
+    window.open(redirectUri, '_self');
```

Comment:
wondering the difference to `window.location.assign`

Reply:
After researching, `window.location.assign` is better since many browsers block pop-up windows (due to abuse like pop-up ads), using `window.open` might be blocked by the browser or restricted by the user's browser settings.